### ISSUE TITLE
libc++ hardening mode doesn't detect

### ISSUE THREAD

**Issue by thesamesam** [llvm/llvm-project#128627]
Noticed when reading https://jacko.io/smart_pointers.html.

libc++'s hardening mode doesn't detect the following, even with `-D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG`, while `-D_GLIBCXX_DEBUG` (though not `-D_GLIBCXX_ASSERTIONS` does):

```c++
#include <vector>

int main() {
    std::vector<int> my_vector = {1, 2, 3};
    for (auto element : my_vector) {
        if (element == 2) {
            my_vector.push_back(4);
            /* The next loop iteration reads a dangling pointer. */
        }
    }
}
```

godbolt: https://godbolt.org/z/jKaYPY3vb

**Comment by frederick-vs-ja** [comment]
It seems intended that such mechanism is not included in hardening. Per [P3471R4](https://wg21.link/p3471r4), hardening shouldn't require ABI change, while the requested mechanism requires more data members.

**Comment by thesamesam** [comment]
Ah, that makes sense. Thanks.

**Comment by philnik777** [comment]
I don't think this should be closed. While I agree that the current flags can't make this work, I don't think we catch this in _any_ mode, including with ABI breaks. We can add them, just behind an ABI flag.
